src/components/Form/formFunctions.ts

```diff
@@ -0,0 +1,156 @@
+import React from "react";
+import { booleanGroup } from "@/components/DataInput/inputHandlerFactories";
+import { Database } from "@/database_types_file";
+import {
+    changeEventHandler,
+    selectChangeEventHandler,
+} from "@/components/DataInput/inputHandlerFactories";
+
+type Field = Fields[keyof Fields];
+export type ErrorSetter = (errorKey: string, errorType: Error) => void;
+export type FieldSetter = (fieldKey: string, newFieldValue: Field) => void;
+export type PersonType = Database["public"]["Enums"]["gender"];
+
+export enum Error {
+    initial = "required",
+    none = "",
+    required = "This is a required field.",
+    invalid = "Please enter a valid entry.",
+    submit = "Please ensure all fields have been entered correctly. Required fields are labelled with an asterisk.",
+    external = "Please try again later.",
+}
+
+export const numberRegex = /^\d+$/;
+
+export interface CardProps {
+    formErrors: FormErrors;
+    errorSetter: ErrorSetter;
+    fieldSetter: FieldSetter;
+    fields: Fields;
+}
+
+export interface Person {
+    personType: PersonType;
+    age?: number | null;
+    quantity?: number;
+}
+
+export interface Fields {
+    [fieldKey: string]: any;
+}
+
+export interface FormErrors {
+    [errorKey: string]: Error;
+}
+
+export const setError = (
+    errorSetter: React.Dispatch<React.SetStateAction<FormErrors>>,
+    errorValues: FormErrors
+): ErrorSetter => {
+    return (errorKey, errorType) => {
+        errorSetter({ ...errorValues, [errorKey]: errorType });
+    };
+};
+
+export const setField = <SpecificFields extends Fields>(
+    fieldSetter: React.Dispatch<React.SetStateAction<SpecificFields>>,
+    fieldValues: SpecificFields
+): FieldSetter => {
+    return (fieldKey, newFieldValue) => {
+        fieldSetter({ ...fieldValues, [fieldKey]: newFieldValue });
+    };
+};
+
+const getErrorType = (
+    input: string,
+    required?: boolean,
+    regex?: RegExp,
+    additionalCondition?: (value: string) => boolean
+): Error => {
+    if (required && input === "") {
+        return Error.required;
+    }
+    if (
+        (regex !== undefined && !input.match(regex)) ||
+        (additionalCondition !== undefined && !additionalCondition(input))
```

Comment:
can you use `!regex` and `!additionalCondition`?

Reply:
Would that not mean the opposite? (Since I want it to be `!== undefined` instead of `=== undefined`?

Reply:
very true, can you do `regex` or `!!regex`?

Reply:
talked in person, this is better